### PR TITLE
LibWeb/CSS: Add support for hwb, oklab and oklch color functions

### PR DESCRIPTION
**Pull request by Gingeh**
This PR breaks the original `parse_rgb_or_hsl_color` method into several `parse_*_color` methods for each color function.
It passes all existing ladybird tests (if you want new tests in this PR then I'll need to be told how to add them), and I didn't find any regressions in the web platform tests.

Here's some before and after screenshots from a website that uses `oklch`:
before
after

### CODE REVIEW

**Comment by AtkinsSJ** [comment]
Hi there, welcome to Ladybird! :^)

I don't have time to properly look at this right now, so I'll come back to it soon. But yes, new tests would be appreciated! If you look through [`Tests/LibWeb`](https://github.com/LadybirdBrowser/ladybird/tree/master/Tests/LibWeb) you'll get an idea of what tests look like. I've made a note to add some documentation for this (#398) because we don't have any and people keep asking. :sweat_smile: 

Specifically because this is testing colours, which are visual, it would make sense as a reference test in the `Ref` subdirectory. Those work by having two different pages, which the test runner loads, takes screenshots of, and makes sure they look identical. So you would:
- Create an html file in `Tests/LibWeb/Ref` called something like `css-color-functions.html`, that uses those color functions.
- Then create a "reference" html file in `Tests/LibWeb/Ref/reference` - by convention the reference for `foo.html` is `foo-ref.html`. This would look identical to your test, but use more basic colour definitions, like a hex colour.
- Inside your first test file, add `<link rel="match" href="reference/ref-test-file-name.html" />` to the `<head>`, with the name of your reference file. That tells the test runner where to look.

Feel free to ask if you have any more questions. :^)

**Comment by Gingeh** [comment]
I've added some tests now (and they helped catch a bug!). There seems to be some kind of accumulated rounding error in a few cases (marked in the reference file) although they're all within 1/256th of the expected rgb values.

**Comment by circl-lastname** [comment]
You need to run some command to generate the tests `expected` files, but as this is a Ref test I have no idea which one. A maintainer will chime in on that probably

**Comment by Gingeh** [comment]
Do I? The test seems to run correctly (and correctly caught an error at first).

**Comment by circl-lastname** [comment]
Sorry I was mistaken how a Ref test worked

File: Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp [hunk 1]
```diff
@@ -2561,6 +2561,232 @@ Optional<Color> Parser::parse_hsl_color(Vector<ComponentValue> const& component_
...
+
+    return Color::from_oklab(L_val, a_val, b_val, alpha_val);
+}
+
+Optional<Color> Parser::parse_oklch_color(Vector<ComponentValue> const& component_values)
+{
+    float L_val;
+    float c_val;
+    float h_val;
+    float alpha_val;
```
Comment (ADKaster) [inline Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2717]: even though they are overwritten later, primitve types should always be initialized (0.0 should work). If you really want an uninitalized value, `Optional<float>` would be preferred to `float foo;`

Comment (Gingeh) [reply Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2717]: yeah that's fair, and I'll move the alpha declaration to later in the functions where it's actually used

**Review by ADKaster — APPROVED** [review]

**Comment by nico** [comment]
This caused the following diags on serenity's CI:

```
/home/runner/work/serenity/serenity/Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2633:23: error: implicit conversion from 'float' to 'double' to match other operand of binary expression [-Werror=double-promotion]
 2633 |     if (w_val + b_val >= 1.0) {
      |         ~~~~~~~~~~~~~~^~~~~~
/home/runner/work/serenity/serenity/Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp: In member function 'AK::Optional<Gfx::Color> Web::CSS::Parser::Parser::parse_oklch_color(const AK::Vector<Web::CSS::Parser::ComponentValue>&)':
/home/runner/work/serenity/serenity/Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:2745:44: error: implicit conversion from 'float' to 'double' to match other operand of binary expression [-Werror=double-promotion]
 2745 |         h_val = hue.token().number_value() * AK::Pi<float> / 180;
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
```

I'm mentioning this in case you're not intentionally using more permissive diags in ladybird.

**Comment by nico** [comment]
This fixes it, fwiw:

```diff
% git diff
diff --git a/Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp b/Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp
index 4c67c4193c..8f31d12349 100644
--- a/Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp
+++ b/Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp
@@ -2630,7 +2630,7 @@ Optional<Color> Parser::parse_hwb_color(Vector<ComponentValue> const& component_
             return {}; // should have consumed all arguments.
     }
 
-    if (w_val + b_val >= 1.0) {
+    if (w_val + b_val >= 1.0f) {
         u8 gray = clamp(llroundf((w_val / (w_val + b_val)) * 255), 0, 255);
         return Color(gray, gray, gray, clamp(llroundf(alpha_val * 255), 0, 255));
     }
@@ -2742,7 +2742,7 @@ Optional<Color> Parser::parse_oklch_color(Vector<ComponentValue> const& componen
         return {};
 
     if (hue.is(Token::Type::Number)) {
-        h_val = hue.token().number_value() * AK::Pi<float> / 180;
+        h_val = static_cast<float>(hue.token().number_value()) * AK::Pi<float> / 180;
     } else {
         auto numeric_value = hue.token().dimension_value();
         auto unit_string = hue.token().dimension_unit();
```

**Pull request merged**